### PR TITLE
Problem Set 2

### PR DESCRIPTION
**Pull request by hannako**
Hi Hywel and Gareth.

There are three problems in this set, Initializing, Caesar and Vigenere. Each is in a separate file and the link to the questions is [here](http://cdn.cs50.net/2016/x/psets/2/pset2/pset2.html).

I am slightly cringing to submit this as my code is horrible!! I know there will be a simpler more elegant solution but I think I approached it wrong. I was abstracting and testing as I went to keep it readable but then I found I'd moved so far from the neat little ci = (pi + k) % 26. I'd spent a lot of time so thought I better move on...Sorry!!

I've not extracted to a module before or used minitest so I've spent a bit longer on this set to read around these things which has been useful.

Jess

### CODE REVIEW

File: p_sets/2/initializing.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+# INITIALIZING
+# You may assume that the users input will contain only letters
+# (uppercase and/or lowercase) plus single spaces between words.
+def ask_name
+  puts "What's your name?"
+  gets.chomp
+end
+
+def initilize_name
```
Comment (hcarver) [inline p_sets/2/initializing.rb]: Typo here - it's missing an "a"

File: p_sets/2/initializing.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  initials = ''
+  ask_name.split.each do |word|
+    initials += word.chr.upcase
+  end
+  initials
```
Comment (hcarver) [inline p_sets/2/initializing.rb]: I think this function could be replaced with a `map` - the clue is that the output of the function has a single item in it for each word in the name. Then you can join them all together to make the initials. That'd be my preferred way.

File: p_sets/2/initializing.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+end
+
+def initial_printer
+  puts format('%s',initilize_name)
+end
```
Comment (hcarver) [inline p_sets/2/initializing.rb:19]: Similar to in the last review, I think it'd be clearer with multiple steps in the middle:
```
name = ask_name
initials = initialize_name(name)
puts initials
```

Because you're getting a string back from `initialize_name`, you don't need to format that, you can just print it directly.

File: p_sets/2/caesar.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def run
+    integer = ARGV.first.to_i
+    if integer > 0
+      puts 'text?'
+      plaintext = STDIN.gets.chomp
+      cyphertext_reporter(plaintext, integer)
+    else
+      puts 1.to_s
```
Comment (hcarver) [inline p_sets/2/caesar.rb]: More for info than a change that needs making, but outputting 1 isn't the same thing as returning 1. Kind of like returning 1 from a function is different to `puts 1`, the same is true for programs as a whole. I think the ruby way of returning 1 from your program is `Kernel.exit(1)` or `Kernel.abort("Positive integer arguments only, please")`

File: p_sets/2/encryption.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
+module Encryption
+
+  ASCII_LETTERS = {
+    capitals: (65..90),
+    lowercase: (97..122)
+  }
+
+  def cyphertext_reporter(plaintext, integer)
+    puts format('%s', encrypt(plaintext, integer))
```
Comment (hcarver) [inline p_sets/2/encryption.rb]: Maybe practising `format` is part of the exercise here, but you can just do `puts encrypt...`

File: p_sets/2/caesar.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def make_encrypted_ascii(plaintext, key)
+    ascii = []
+    plaintext.each_char do |character|
+      if letter?(character)
+        ascii << apply_key(character, key)
+      else ascii << character
+      end
+    end
```
Comment (hcarver) [inline p_sets/2/caesar.rb:31]: Whenever there's one item in the output for each item in the input, you can use a `map`, which is more ruby-ish and (I'd have thought) faster.

I think our coding standards ask for a new line after the `else` before the `ascii << character`.

Comment (hcarver) [reply p_sets/2/caesar.rb:31]: It also means you don't need to declare a new variable to store the result, or return that variable at the end, which saves 2 lines of code 👍

File: p_sets/2/caesar_test.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    assert_equal  'BcB', @caesar.encrypt('AbA', 1),
+                  'encryption method maintains case of letters'
+    assert_equal  'Bc_-!@£$%^&*()B123', @caesar.encrypt('Ab_-!@£$%^&*()A123', 1),
+                  'encryption ignores non-alphanumerics'
+    assert_equal  'bcde', @caesar.encrypt('wxyz', 5),
+                  'encryption works at end of alphabet'
+    assert_equal  'B^^123e', @caesar.encrypt('W^^123z', 5),
+                  'encryption works at end of alphabet'
+    assert_equal 'abc', @caesar.encrypt('zab', 27),
+                  'accepts integers greater than 26'
```
Comment (hcarver) [inline p_sets/2/caesar_test.rb:23]: Good testing, you've obviously thought about all the edge cases. 👍

File: p_sets/2/encryption.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def encrypt(plaintext, string)
```
Comment (hcarver) [inline p_sets/2/encryption.rb]: This method gets called with an arg called `integer` but here it's called `string`. Neither of those are ideal, IMO. I think they're both `key` or `shift_key` aren't they?

File: p_sets/2/encryption.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    key = make_key(string)
+    encrypted_ascii = make_encrypted_ascii(plaintext, key)
+    make_cyphertext(encrypted_ascii)
+  end
+
+  def apply_key(character, key)
+    capital_letter?(character) ? size = :capitals : size = :lowercase
+    ascii = character.ord + key
+    ascii > ASCII_LETTERS[size].max ? ascii - 26 : ascii
```
Comment (hcarver) [inline p_sets/2/encryption.rb]: I feel like there should be a neater way of writing this, but I couldn't do better than 
`((key + character.ord - ASCII_LETTERS[size].min) % 26) + ASCII_LETTERS[size].min`
which isn't necessarily an improvement.

File: p_sets/2/encryption.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def make_cyphertext(ascii)
+    cyphertext = []
+    ascii.each do |integer|
+      cyphertext << integer.chr
+    end
```
Comment (hcarver) [inline p_sets/2/encryption.rb]: Another case for a map. In fact, ruby makes it v easy to map to a function of each element: you can do `ascii.map(&:chr).join` for the whole function.

File: p_sets/2/vigenere.rb [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+# to the next character in k.
+require_relative 'encryption'
+class Vigenere
+  include Encryption
+
+  def make_key(string)
+    key = []
+    string.upcase.each_char do |letter|
+      key << letter.ord - ASCII_LETTERS[:capitals].min
+    end
```
Comment (hcarver) [inline p_sets/2/vigenere.rb:20]: Another case for a `map`. It's not a very C++ concept, so it might not be covered in the course.

File: p_sets/2/vigenere.rb [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    key
+  end
+
+  def apply_key(character, key, index)
+    capital_letter?(character) ? size = :capitals : size = :lowercase
+    ascii = character.ord + key[index].to_i
+    ascii > ASCII_LETTERS[size].max ? ascii - 26 : ascii
```
Comment (hcarver) [inline p_sets/2/vigenere.rb:30]: You could, if you wanted, abstract the second two lines of this to an `apply_shift` method in the Encryption module, and reuse it from both here and the Caesar cipher file. `apply_shift(letter, shift)` and then each of the two ciphers can work out what it's on shift amount is (easy for Caesar) and call that method.

File: p_sets/2/vigenere.rb [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  def make_encrypted_ascii(plaintext, key)
+    ascii = []
+    index = 0
+    plaintext.each_char do |character|
+      if letter?(character)
+        ascii << apply_key(character, key, index)
+        index < key.length - 1 ? index += 1 : index = 0
```
Comment (hcarver) [inline p_sets/2/vigenere.rb]: I would find `index = (index + 1) % key.length` neater, personally. I don't have a strong argument for why I prefer that though. Also, note this function could use a `map` again.

File: p_sets/2/vigenere.rb [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+      else
+        ascii << character
+      end
+    end
+    ascii
+  end
+
+  def valid?(string)
+    string.index(/[^a-zA-Z]/).nil?
```
Comment (hcarver) [inline p_sets/2/vigenere.rb]: Nice! I think there might be a `matches?` method that does something similar to this - which would be a more readable expression of what you're doing - `!string.matches(/regex/)`

File: p_sets/2/vigenere.rb [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  def run
+    string = ARGV.first
+    if string.nil? || !ARGV.count == 1 || !valid?(string)
+      puts 1.to_s
+    else
+      puts "text?"
+      plaintext = STDIN.gets.chomp
+      cyphertext_reporter(plaintext, string)
```
Comment (hcarver) [inline p_sets/2/vigenere.rb]: Again, I'd prefer to split this second step into something like
```
key = ....
cypher_text = encode(plaintext, key)
puts cypher_text
```

File: test/reports/TEST-CaesarProgramTest.xml [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<testsuite time='0.026635' skipped='0' failures='1' errors='0' name="CaesarProgramTest" assertions='4' tests='1'>
+  <testcase time='0.026635' file="p_sets/caesar_test.rb" name="test_encrypt" assertions='4'>
+    <failure type='Minitest::Assertion' message="encryption works at end of alphabet.&#13;&#10;--- expected&#13;&#10;+++ actual&#13;&#10;@@ -1 +1,2 @@&#13;&#10;-&quot;bcde&quot;&#13;&#10;+# encoding: US-ASCII&#13;&#10;+&quot;|}~\\x7F&quot;&#13;&#10;">p_sets/caesar_test.rb:21:in `test_encrypt&#39;
+    </failure>
+  </testcase>
+</testsuite>
```
Comment (hcarver) [inline test/reports/TEST-CaesarProgramTest.xml:5]: Probably didn't mean to commit this test report. You might want to add `test/reports` to .gitignore to avoid seeing these.

**Review by hcarver — CHANGES_REQUESTED** [review]

File: p_sets/2/encryption.rb (hunk 8 above)
Comment (hannako) [reply p_sets/2/encryption.rb]: I opted for string and integer because I have two make_key methods that produce workable keys from the user inputs. But from a user pov i see that their inputs are the keys, so i should rename the variables passed from make_key to the encryption methods maybe?

File: p_sets/2/initializing.rb (hunk 3 above)
Comment (hannako) [reply p_sets/2/initializing.rb:19]: Yes sorry for making the same mistake again! Makes sense.

File: p_sets/2/caesar.rb (hunk 6 above)
Comment (hannako) [reply p_sets/2/caesar.rb:31]: 👍

File: p_sets/2/encryption.rb (hunk 5 above)
Comment (hannako) [reply p_sets/2/encryption.rb]: No i was just using it because I learnt it last week from Gareth's review. I'm not sure when to use it and when not. I'm going to read up on it.

File: p_sets/2/encryption.rb (hunk 9 above)
Comment (hannako) [reply p_sets/2/encryption.rb]: I had the version you've suggested :) pleased I was on the right track! I found it hard to follow once I'd worked it out so I went back to a long clunky one. I'll switch back

File: p_sets/2/encryption.rb (hunk 10 above)
Comment (hannako) [reply p_sets/2/encryption.rb]: Thank you

File: p_sets/2/caesar.rb [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+# Otherwise prompt the user for a string of plaintext and then output ciphertex.
+# non-alphabetical characters should be outputted unchanged.
+# After outputting this ciphertext, your program should exit, returning 0.
+# you may not assume that k will be less than or equal to 26
+#  should work for all non-negative ints less than 2v31 - 26
+require_relative 'encryption'
+class Caesar
+  include Encryption
+
+  def make_key(integer)
```
Comment (gareth) [inline p_sets/2/caesar.rb]: Since this method has no side effects, I probably wouldn't have the `make_` prefix on the name. I can't find a simple, consistent way to phrase it, but generally you don't have to explain that a method is `calculate_`ing or `get_`ting or `make_`ing something - that's implied from the fact that you're calling a method.

File: p_sets/2/caesar.rb (hunk 4 above)
Comment (gareth) [reply p_sets/2/caesar.rb]: I agree with this, but I'll add that it's better design if the only place that cares about the exit status is the outer level of your app.

Whereas C has `main` as a separate function, in Ruby you can think of the file itself as being the `main` function - which is why line 42 below works.

So with that in mind, I'd probably have `run` return a value on success, and `nil` on failure, and then use that in the main execution to decide how to exit the program:

@gareth Do you mean like this...

```def run
    integer = ARGV.first.to_i
    if integer > 0
      puts 'text?'
      plaintext = STDIN.gets.chomp
      cyphertext_reporter(plaintext, integer)
    else
      false
    end
  end
```
@gareth How do I know what the exit codes are without using Kernel.abort? I can see how to get the exit codes when calling a method in bash. But not when running a program from the command line. 

```
if Caesar.new.run
  exit 0
else
  exit 1
end
```

There are other changes I might make to improve encapsulation here, but this is the main one.

File: p_sets/2/encryption.rb [hunk 18]
```diff
@@ -0,0 +1,43 @@
+module Encryption
+
+  ASCII_LETTERS = {
+    capitals: (65..90),
+    lowercase: (97..122)
```
Comment (gareth) [inline p_sets/2/encryption.rb:4]: Ruby actually allows you to have ranges of pretty much anything (well, anything that has a definitive ordered enumerable state).

It means that you can do this:
```
uppercase = "A".."Z"
uppercase.cover? "F" # => true
uppercase.cover? "x" # => false
uppercase.cover? "!" # => false
```

This is mostly informative, because I know the C course spoke a lot about the ASCII values of characters, but it's definitely possible to hide a lot of that in Ruby. So I know this simplifies your `cover?` use below, but it probably makes some of the rest of the code more complicated.

File: p_sets/2/encryption.rb (hunk 5 above)
Comment (gareth) [reply p_sets/2/encryption.rb]: I use `format` when I'm combining strings with other characters that might otherwise get lost in the string interpolation syntax.
```
"#{foo} #{bar}" # OK
"Result for #{foo}: #{bar}" # Less good
format("Result for %s: %s", foo, bar) # Clearer
```

It's still quite subjective, but one advantage of being able to separate the format from the values is they can come from different places - the format string might be a constant, the values might be method arguments.

File: p_sets/2/caesar.rb (hunk 17 above)
Comment (gareth) [reply p_sets/2/caesar.rb]: Also, to pick up Hywel's point about argument names, `integer` is possible a better name than `x` would be, but still can be improved. To use this class I'd need to know what the integer represents.

In this case, it looks like the user will be passing in a number that they intend to use as a key, and you're constraining it to the range `0..25`

The most explicit you could be in this case is probably: `def valid_key(intended_key)`.

You could shorten that either to `def key(intended_key)` or `def valid_key(key)`, but `def key(key)` would remove too much context I think.

File: p_sets/2/encryption.rb [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def encrypt(plaintext, string)
+    key = make_key(string)
+    encrypted_ascii = make_encrypted_ascii(plaintext, key)
+    make_cyphertext(encrypted_ascii)
+  end
+
+  def apply_key(character, key)
+    capital_letter?(character) ? size = :capitals : size = :lowercase
```
Comment (gareth) [inline p_sets/2/encryption.rb]:
```
size = capital_letter?(character) ? :capitals : :lowercase
```

**Review by gareth — COMMENTED** [review]
Hywel's picked up on a lot, so I've added some style comments

File: p_sets/2/caesar.rb (hunk 4 above)
Comment (hannako) [reply p_sets/2/caesar.rb]: @gareth I'm not sure if you get alerts when I @ you inside existing comments, so pinging you here just in case you didn't see. Not in hassle-y way.
